---
output:
  html_document:
    toc: true
    toc_float: false
  pdf_document:
    toc: true
params:
  path_prefix:
    value: ""
  vector_type:
    value: ""
  target_start:
    value: 0
  target_end:
    value: 0
  target_start_repcap:
    value: 0
  target_end_repcap:
    value: 0
---

```{r setup, include=FALSE}
library(tidyverse)
library(flextable)

# Avoid scientific notation
options(scipen = 999)
# Hush a benign message about "grouped output"
options(dplyr.summarise.inform = FALSE)

# Configure Rmarkdown/knitr/pandoc processing
knitr::opts_chunk$set(echo = FALSE,
                      fig.align = "center",
                      out.width = "80%")
# Global ggplot2 theme
theme_set(theme_minimal())
# Table style and theme
set_flextable_defaults(
  font.family = "Verdana",
  padding = 5,
  tabcolsep = 5,
  table.layout = "autofit",
  theme_fun = "theme_vanilla"
)
# Load sample metadata
meta <- read_tsv(paste0(params$path_prefix, '.metadata.tsv'), show_col_types = FALSE)
```

---
title: "AAV Sequence Analysis"
subtitle: 'Sample name: `r meta$sample_display_name`'
date: 'Date: `r format(Sys.Date())`'
---

# Analysis context

|                       | Value                      |
|:---------------------:|----------------------------|
| Sample unique ID      | `r meta$sample_unique_id`  |
| Sequencing run ID     | `r meta$sequencing_run_id` |
| Construct vector type | `r params$vector_type`AAV  |


`\newpage`{=latex}


# Read-based AAV vector type classification

## Definitions

Table: Reference label definitions.

Reference Label    |Definition
:-----------------:|----------------------------------------------------
vector             |Read originates from the vector plasmid.
repcap             |Read originates from the RepCap plasmid. The Rep gene encodes four proteins (Rep78, Rep68, Rep52, and Rep40), which are required for viral genome replication and packaging, while Cap expression gives rise to the viral capsid proteins (VP; VP1/VP2/VP3), which form the outer capsid shell that protects the viral genome, as well as being actively involved in cell binding and internalization.
helper             |Read originates from the helper plasmid. In addition to Rep and Cap, AAV requires a helper plasmid containing genes from adenovirus. These genes (E4, E2a and VA) mediate AAV replication.
host               |Read originates from the host genome that is given (e.g. hg38, CHM13).
chimeric-vector    |One part of the read aligns to the vector plasmid, while another part of the same read aligns to a different reference sequence.
chimeric-nonvector |The read consists of fragments that align to two or more different reference sequences, neither of which is the vector plasmid.


Table: Assigned type definitions.

Assigned Type|Definition
:-----------:|----------------------------------------------------
ssAAV        |Single-stranded AAV vector genome where the resulting DNA template is expected to be single-stranded, as opposed to self-complementary. A sequencing read is inferred as ssAAV if it has a single alignment to the vector plasmid’s ITR-to-ITR payload region and no complementary supplemental alignment.
scAAV        |Self-complementary AAV vector genome where one half of the payload region is a reverse complement of the other, resulting in an intramolecular double-stranded DNA template. A sequencing read is inferred as scAAV if it aligns to the payload region in both forward (+) and reverse (-) read directions.
backbone     |Read aligns to the vector plasmid sequence but fully outside of the annotated ITR-to-ITR region, indicating that the sequence fragment originated solely from the plasmid backbone.
other-vector |Read consists of a fragment mapping to the vector but with characteristics other than those listed above.


`<br />`{=html}

![Single-stranded AAV types (ssAAV).
ssAAV reads are expected to have a single (forward) alignment to the vector genome.
"ssAAV full" reads cover the target region from left ITR to right ITR.
"ssAAV left-partial" contain the left ITR but are missing the right ITR.
"scAAV vector+backbone" reads overlap both the ITR-to-ITR region and the plasmid backbone, beyond the ITR.
](class_def_ssAAV.png)

`<br />`{=html}

![Self-complementary AAV types (scAAV).
scAAV reads are expected to have both forward and reverse-complementary alignments to the vector genome.
The definitions of full, partial, and backbone reads are the same as in ssAAV.
](class_def_scAAV.png)

`\break`{=latex}


```{r calculate, message=FALSE, warning=FALSE}
# -------------------------------------
# alignments.tsv.gz
# -------------------------------------

x.all.summary <- read_tsv(paste0(params$path_prefix, '.alignments.tsv.gz'), show_col_types = FALSE)


# ---------------------------------------
# per_read.tsv.gz
# ---------------------------------------

x.all.read <- read_tsv(paste0(params$path_prefix, '.per_read.tsv.gz'), show_col_types = FALSE)

# Set the display order for categoricals
valid_types <- c('ssAAV', 'scAAV', 'other-vector', 'backbone')
valid_subtypes <- c(
  'full',
  'full-gap',
  'read-through',
  'partial',
  'left-partial',
  'right-partial',
  'itr-partial',
  'left-snapback',
  'right-snapback',
  'snapback',
  'unresolved-dimer',
  'tandem',
  'complex',
  'unclassified',
  'backbone')

ref_counts <- x.all.read %>%
  filter(reference_label != "(unmapped)") %>%
  group_by(reference_label) %>%
  summarise(e_count = sum(effective_count)) %>%
  arrange(desc(e_count))
ref_levels <- c(ref_counts$reference_label, "(unmapped)")

x.all.read <- x.all.read %>%
  mutate(
    reference_label = factor(reference_label, levels = ref_levels),
    assigned_type = factor(assigned_type, levels = valid_types),
    assigned_subtype = factor(assigned_subtype, levels = valid_subtypes))

total_read_count_all <- sum(x.all.read$effective_count)

# Filter to vector-only and sort types/subtypes
x.read.vector <- x.all.read %>% filter(reference_label == "vector")

total_read_count_vector <- sum(x.read.vector$effective_count)

x.summary.primary <- filter(x.all.summary, is_mapped == "Y", is_supp == "N")
x.joined.vector <- left_join(x.read.vector, x.summary.primary,
                             by = "read_id", multiple = "first") %>%
                    uncount(effective_count)

# -------------------------------------
# Aggregate tables
# -------------------------------------

agg_ref_type <- read_tsv(paste0(params$path_prefix, '.agg_ref_type.tsv'), show_col_types = FALSE) %>%
  mutate(
    reference_label = factor(reference_label, levels = ref_levels),
    assigned_type = factor(assigned_type, levels = valid_types)) %>%
  arrange(reference_label)

agg_subtype <- read_tsv(paste0(params$path_prefix, '.agg_subtype.tsv'), show_col_types = FALSE) %>%
  mutate(
    assigned_type = factor(assigned_type, levels = valid_types),
    assigned_subtype = factor(assigned_subtype, levels = valid_subtypes))
```

# Assigned types by read alignment characteristics

```{r typetable, message=FALSE, warning=FALSE}
flextable(agg_ref_type %>% select(! pct_vector)) %>%
  set_header_labels(values = c("Mapped Reference", "Assigned Type", "Count", "Frequency (%)"))
```

```{r typeplots, message=FALSE, warning=FALSE}
countreads <- agg_ref_type %>% filter(reference_label != "lambda")

# Bar chart
ggplot(countreads, aes(x = reference_label, y = pct_total, fill = assigned_type)) +
  geom_bar(stat = "identity") +
  scale_x_discrete("Sequence", guide = guide_axis(angle = 30)) +
  scale_y_continuous("Reads (%)",
                     limits = c(0, 100),
                     expand = c(0, 0))

# Pie chart
ggplot(countreads, aes(x = "", y = effective_count, fill = reference_label)) +
  geom_bar(stat = "identity", width = 1) +
  coord_polar("y", start = 0) +
  theme_void()
```


## Frequency of vector assigned types

```{r vectypes, message=FALSE, warning=FALSE}
vectypes <- agg_ref_type %>%
    filter(reference_label == "vector") %>%
    select(! reference_label)

flextable(vectypes) %>%
  set_header_labels(values = c(
    "Assigned Type",
    "Count",
    "Frequency in AAV (%)",
    "Total Frequency (%)"
  ))
```

# Distribution of vector read lengths by assigned types

```{r atypeviolin, message=FALSE, warning=FALSE}
p2.atype_violin <- ggplot(
  x.joined.vector %>%
    mutate(assigned_type = factor(assigned_type, levels = valid_types)),
  aes(x = assigned_type, y = read_len)
) +
  geom_violin() +
  labs(title = "Distribution of vector read lengths by assigned type",
       x = "Assigned type") +
  scale_y_continuous("Read length",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05)))

p2.atype_violin
```


# Assigned subtypes detailed analysis

## Definitions

```{asis, echo = params$vector_type == "ss"}
Table: Assigned subtype definitions for single-stranded input vector.

Assigned Type|Assigned Subtype|Definition
:-----------:|:--------------:|------------------------------------------------
ssAAV        |full            |Read aligns to a fragment of the vector originating from the left (upstream) ITR and ending at the right (downstream) ITR of the vector.
ssAAV        |full-gap        |Read aligns to the vector ITR-to-ITR region, as with “full”, but with a significant number of gaps in the alignment between the ITRs.
ssAAV        |read-through    |Read aligns to a fragment including the vector as well as plasmid backbone sequence. May imply read-through beyond the right ITR, or reverse packaging if the alignment is to only the left ITR and backbone.
ssAAV        |partial         |Read aligns to a fragment of the vector originating from within the ITR sequences.
ssAAV        |left-partial    |Read aligns to a fragment of the vector originating from the left (upstream) ITR of the vector while not covering the right ITR.
ssAAV        |right-partial   |Read aligns to a fragment of the vector originating from the right (downstream) ITR of the vector while not covering the left ITR.
ssAAV        |left-snapback   |Read consists of a double-stranded, sub-genomic fragment including only the left ITR and aligned symmetrically to the (+) and (-) strands.
ssAAV        |right-snapback  |Read consists of a double-stranded, sub-genomic fragment including only the right ITR and aligned symmetrically to the (+) and (-) strands.
other-vector |snapback        |Read aligns to a double-stranded fragment in both (+) and (-) strands, but does not include either ITR.
other-vector |unresolved-dimer|Read aligns to a double-stranded fragment covering the full ITR-to-ITR region in both (+) and (-) strands. A dimer in ssAAV context, twice the size of a ssAAVV-full vector genome.
other-vector |tandem          |Read has two or more overlapping alignments on the same strand, but none on the reverse strand, indicating tandem duplication of the same region.
other-vector |complex         |Read aligns to a double-stranded fragment with asymmetrical and/or multiple alignments on the (+) and (-) strands.
other-vector |unclassified    |Read alignment does not match any of the above orientations.
```

```{asis, echo = params$vector_type == "sc"}
Table: Assigned subtype definitions for self-complementary input vector.

Assigned Type|Assigned Subtype|Definition
:-----------:|:--------------:|------------------------------------------------
scAAV        |full            |Read aligns to a fragment of the vector originating from the left (upstream) ITR and ending at the right (downstream) ITR of the vector.
scAAV        |full-gap        |Read aligns to the vector ITR-to-ITR region, as with “full”, but with a significant number of gaps in the alignment between the ITRs.
scAAV        |read-through    |Read aligns to a fragment including the vector as well as plasmid backbone sequence. May imply read-through beyond the right ITR, or reverse packaging if the alignment is to only the left ITR and backbone.
scAAV        |partial         |Read aligns to a fragment of the vector originating from within the ITR sequences.
scAAV        |itr-partial     |Read aligns only to a single-stranded fragment of the vector originating from the left ITR while not covering the right ITR.
scAAV        |snapback        |Read consists of a double-stranded, sub-genomic fragment including only the left ITR and read alignments in both (+) and (-) strands.
other-vector |tandem          |Read has two or more overlapping alignments on the same strand, but none on the reverse strand, indicating tandem duplication of the same region.
other-vector |complex         |Read aligns to a double-stranded fragment with asymmetrical and/or multiple alignments on the (+) and (-) strands.
other-vector |unclassified    |Read alignment does not match any of the above orientations.
```

## Assigned types and subtypes

```{r subtypetable, message=FALSE, warning=FALSE}
flextable(agg_subtype) %>%
  set_header_labels(
    values = c(
      "Assigned Type",
      "Assigned Subtype",
      "Count",
      "Freq. in AAV (%)",
      "Total Freq. (%)"
    )
  )
```


## Distribution of read length by subtype

```{r lenhisto2_subtype, message=FALSE, warning=FALSE}
# Single plot
p1_subtypes_len_hist <- ggplot(filter(x.joined.vector),
                               aes(x = read_len, color = assigned_subtype)) +
  geom_freqpoly(binwidth = 50) +
  labs(title = "Distribution of read length by subtype") +
  scale_x_continuous("Read length (bp)",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05))) +
  scale_y_continuous("Count (binned)",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05)))
p1_subtypes_len_hist
```


# AAV mapping to reference sequence

## Gene therapy construct

```{r lenhisto2_construct, message=FALSE, warning=FALSE}
p1.map_starts <- ggplot(x.joined.vector, aes(map_start0 + 1, fill = assigned_subtype)) +
  geom_histogram(aes(y = after_stat(count) / sum(after_stat(count))),
                 binwidth = 125,
                 boundary = 0,
                 closed = "left") +
  geom_vline(xintercept = params$target_start,
             color = "black",
             lty = 2) +
  geom_vline(xintercept = params$target_end,
             color = "black",
             lty = 2) +
  scale_x_continuous("Mapped reference start position",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05))) +
  scale_y_continuous("Fraction of reads",
                     limits = c(0, 1),
                     expand = c(0, 0)) +
  labs(title = "Distribution of mapped read starts on construct")

p1.map_ends <- ggplot(x.joined.vector, aes(map_end1, fill = assigned_subtype)) +
  geom_histogram(aes(y = after_stat(count) / sum(after_stat(count))),
                 binwidth = 125,
                 boundary = 0,
                 closed = "left") +
  geom_vline(xintercept = params$target_start,
             color = "black",
             lty = 2) +
  geom_vline(xintercept = params$target_end,
             color = "black",
             lty = 2) +
  scale_x_continuous("Mapped reference end position",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05))) +
  scale_y_continuous("Fraction of reads",
                     limits = c(0, 1),
                     expand = c(0, 0)) +
  labs(title = "Distribution of mapped read ends on construct")

p1.map_len <- ggplot(x.joined.vector, aes(map_len, fill = assigned_subtype)) +
  geom_histogram(aes(y = after_stat(count) / sum(after_stat(count))),
                 binwidth = 125,
                 boundary = 0,
                 closed = "left") +
  scale_x_continuous("Mapped reference length",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05))) +
  scale_y_continuous("Fraction of reads",
                     limits = c(0, 1),
                     expand = c(0, 0)) +
  labs(title = "Distribution of mapped spanning region sizes on construct")

p1.map_starts
p1.map_ends
p1.map_len
```

<!-- optional RepCap section -->

```{r lenhisto2_rc, message=FALSE, warning=FALSE, results='asis'}
x.read.repcap <- filter(x.all.read, reference_label == 'repcap')

# Only render this subsection if enough reads mapped to repcap
if (dim(x.read.repcap)[1] > 10) {
  x.summary.repcap <- filter(x.all.summary, read_id %in% x.read.repcap$read_id)

  # Subsection heading in Markdown
  cat("## RepCap\n\n")

  p1.map_starts.repcap <- ggplot(x.summary.repcap, aes(map_start0 + 1)) +
    geom_histogram(aes(y = after_stat(count) / sum(after_stat(count))),
                   binwidth = 125,
                   boundary = 0,
                   closed = "left") +
    geom_vline(xintercept = params$target_start_repcap,
               color = 'red',
               lty = 2) +
    labs(title = "Distribution of mapped read starts on RepCap") +
    scale_x_continuous("Mapped reference start position",
                       limits = c(0, NA),
                       expand = expansion(mult = c(0, 0.05))) +
    scale_y_continuous("Fraction of reads",
                       limits = c(0, NA),
                       expand = expansion(mult = c(0, 0.05)))

  p1.map_starts.repcap
}
```

```{r lenhisto2_rc_end, message=FALSE, warning=FALSE, results='asis'}
if (exists("x.summary.repcap")) {
  p1.map_ends.repcap <- ggplot(x.summary.repcap, aes(map_end1)) +
    geom_histogram(aes(y = after_stat(count) / sum(after_stat(count))),
                   binwidth = 125,
                   boundary = 0,
                   closed = "left") +
    geom_vline(xintercept = params$target_end_repcap,
               color = 'red',
               lty = 2) +
    labs(title = "Distribution of mapped read ends on RepCap") +
    scale_x_continuous("Mapped reference end position",
                       limits = c(0, NA),
                       expand = expansion(mult = c(0, 0.05))) +
    scale_y_continuous("Fraction of reads",
                       limits = c(0, NA),
                       expand = expansion(mult = c(0, 0.05)))

  p1.map_ends.repcap
}
```

```{r lenhisto2_rc_len, message=FALSE, warning=FALSE, results='asis'}
if (exists("x.summary.repcap")) {
  p1.map_len.repcap <- ggplot(x.summary.repcap, aes(map_len)) +
    geom_histogram(aes(y = after_stat(count) / sum(after_stat(count))),
                   binwidth = 125,
                   boundary = 0,
                   closed = "left") +
    labs(title = "Distribution of mapped spanning region sizes on RepCap") +
    scale_x_continuous("Mapped reference length",
                       limits = c(0, NA),
                       expand = expansion(mult = c(0, 0.05))) +
    scale_y_continuous("Fraction of reads",
                       limits = c(0, NA),
                       expand = expansion(mult = c(0, 0.05)))

  p1.map_len.repcap
}
```


# Distribution of non-matches by reference position

```{r varcalc, message=FALSE, warning=FALSE}
# --------------------------------------------------
# nonmatch.tsv.gz
# --------------------------------------------------

x.all.err <- read_tsv(paste0(params$path_prefix, '.nonmatch.tsv.gz'), show_col_types = FALSE)

# Filter for ss/scAAV vector only
x.err.vector <- filter(x.all.err, read_id %in% x.read.vector$read_id)

# Remove duplicated lines
#x.err.vector <- x.err.vector[!duplicated(x.err.vector), ]
x.err.vector <- distinct(x.err.vector)

# Spell out nonmatch types -- from CIGAR codes to words
x.err.vector[x.err.vector$type == 'D', "type"] <- 'deletion'
x.err.vector[x.err.vector$type == 'I', "type"] <- 'insertion'
x.err.vector[x.err.vector$type == 'X', "type"] <- 'mismatch'
x.err.vector[x.err.vector$type == 'N', "type"] <- 'gaps'

# Round positions down to 10s for plotting
x.err.vector$pos0_div <- (x.err.vector$pos0 %/% 10 * 10)

df.err.vector <- x.err.vector %>%
  group_by(pos0_div, type) %>%
  summarise(count = n())
```

```{r varplots, message=FALSE, warning=FALSE}
p1.err_sub <- ggplot(filter(df.err.vector, type == 'mismatch'),
                     aes(x = pos0_div, y = count)) +
  geom_bar(fill = 'darkgreen', stat = 'identity') +
  xlim(c(params$target_start, params$target_end)) +
  labs(title = "Nucleotide substitutions along vector genome",
       subtitle = "Counts binned by 10 nucleotides (NTs) (y-axis) and position (x-axis)",
       x = "Reference position") +
  scale_y_continuous("Counts",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05)))

p1.err_del <- ggplot(filter(df.err.vector, type == 'deletion'),
                     aes(x = pos0_div, y = count)) +
  geom_bar(fill = 'darkred', stat = 'identity') +
  xlim(c(params$target_start, params$target_end)) +
  labs(title = "Nucleotide deletions along vector genome",
       subtitle = "Counts binned by 10 nucleotides (NTs) (y-axis) and position (x-axis)",
       x = "Reference position") +
  scale_y_continuous("Counts",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05)))

p1.err_ins <- ggplot(filter(df.err.vector, type == 'insertion'),
                     aes(x = pos0_div, y = count)) +
  geom_bar(fill = 'darkblue', stat = 'identity') +
  xlim(c(params$target_start, params$target_end)) +
  labs(title = "Insertions by reference position",
       subtitle = "Counts binned by 10 nucleotides (NTs) (y-axis) and position (x-axis)",
       x = "Reference position") +
  scale_y_continuous("Counts",
                     limits = c(0, NA),
                     expand = expansion(mult = c(0, 0.05)))

p1.err_sub
p1.err_del
p1.err_ins
```


# Flip/flop configurations

Term|Definition
:--:|----------------
Flip/Flop|One ITR is formed by two palindromic arms, called B–B' and C–C', embedded in a larger one, A–A'. The order of these palindromic sequences defines the flip or flop orientation of the ITR. ([Read more](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5655423/))

```{r ffcalc, message=FALSE, warning=FALSE}
# --------------------------------------------------
# flipflop.tsv.gz (if available)
# --------------------------------------------------

agg_flipflop_tsv <- paste0(params$path_prefix, '.agg_flipflop.tsv')
if (file.exists(agg_flipflop_tsv)) {
  df.flipflop <- read_tsv(agg_flipflop_tsv, show_col_types = FALSE)
}
```

```{r ffplot, message=FALSE, warning=FALSE, results='asis'}
if (exists("df.flipflop") && (nrow(df.flipflop) > 1)) {
  # Generate flip/flop summary plot
  ffplot <- df.flipflop %>%
    mutate(
      class = if_else(
        leftITR == 'unclassified' | rightITR == 'unclassified',
        "unclassified",
        paste(leftITR, rightITR, sep = '-')
      )
    ) %>%
    mutate(
      unclass = if_else( class == "unclassified", "unclassified", "classified")
    ) %>%
    group_by(type, subtype, class, unclass) %>%
    summarize(classct = sum(count)) %>%
    filter(subtype == 'full')
  ggplot(ffplot, aes(x = class, y = classct, fill = unclass)) +
    geom_bar(stat = 'identity', show.legend = FALSE) +
    labs(x = 'Assigned type') +
    scale_y_continuous("Number of reads",
                       limits = c(0, NA),
                       expand = expansion(mult = c(0, 0.05)))

} else {
  cat("No flip/flop analysis results available to display.")
}
```

```{r fftable, message=FALSE, warning=FALSE, results='asis'}
if (exists("df.flipflop") && (nrow(df.flipflop) > 1)) {
  flextable(df.flipflop) %>%
    set_header_labels(values = c("type", "subtype", "leftITR", "rightITR", "count"))
}
```

```{r saverdata, message=FALSE, warning=FALSE}
# Enable for troubleshooting -- takes some time to generate
# save.image(file=paste0(params$path_prefix, '.Rdata'))
```

`\break`{=latex}

# Methods

This report was generated by an automated analysis of long-read sequencing data from
adeno-associated virus (AAV) products.
The sequencing data should be from the PacBio sequencer run in AAV mode, or equivalent
circular consensus sequencing (CCS) reads (Travers et al., 2010).

In this analysis, reads are aligned to the given AAV, packaging, and host reference
sequences using Minimap2 (Li, 2018).
The reference sequences for each primary alignment and its orientation are counted and
summarized to assign read type classifications, including vector, non-vector, and
chimeric reads.
For reads assigned to the AAV vector, the primary alignment coordinates are compared to
the annotated vector region in the reference sequence, which comprises the left and
right ITRs and the genomic region between them, to assign each read to a subtype
classification.
Sequence variants relative to the vector reference sequence are determined directly from
each read's alignment, specifically the CIGAR string indicating insertions, deletions,
mismatches, and gaps.


## Citations

1. Travers, K. J., Chin, C.-S., Rank, D. R., Eid, J. S. & Turner, S. W. A flexible and efficient template format for circular consensus sequencing and SNP detection. Nucleic Acids Research 38, e159–e159 (2010).
2. Li, H. Minimap2: Pairwise alignment for nucleotide sequences. Bioinformatics 34, 3094–3100 (2018).
